[tensorexpr] check for index out of bounds in ir_eval

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68858

when executing with ir_eval, check for index out of bounds.

Differential Revision: [D32657881](https://our.internmc.facebook.com/intern/diff/D32657881)

davidberard98 committed:
[WIP][tensorexpr] check for index out of bounds in ir_eval

when executing with ir_eval, check for index out of bounds.

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][tensorexpr] check for index out of bounds in ir_eval"


when executing with ir_eval, check for index out of bounds.

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][tensorexpr] check for index out of bounds in ir_eval"


when executing with ir_eval, check for index out of bounds.

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][tensorexpr] check for index out of bounds in ir_eval"


when executing with ir_eval, check for index out of bounds.

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][tensorexpr] check for index out of bounds in ir_eval"


when executing with ir_eval, check for index out of bounds.

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][tensorexpr] check for index out of bounds in ir_eval"


when executing with ir_eval, check for index out of bounds.

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32657881).

davidberard98 committed:
Update on "[WIP][tensorexpr] check for index out of bounds in ir_eval"


when executing with ir_eval, check for index out of bounds.

Differential Revision: [D32657881](https://our.internmc.facebook.com/intern/diff/D32657881)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32657881).

davidberard98 commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -941,6 +941,7 @@ Tensor TensorExprKernel::bindInput(const torch::jit::Value* input) {
           "t" + input_name_map_[input],
           {0},
```
to handle this case (where buffer size is set to 0) I added the is_flattened_ variable; is there a better way to handle this?

davidberard98 replied:
How about the case of non-contiguous tensors, e.g. 1-dimensional tensors with stride that is not 1? In that case, how could we tell that we are handling the special case of a 1d flattened buffer?

huiguoo replied:
If I remember correctly, NNC will convert dis-contiguous tensors to contiguous buffers before generating code. That is, all buffers inside NNC are contiguous but this line : https://github.com/pytorch/pytorch/blob/3b27304d20f53619344b74e683357424ca3d4029/torch/csrc/jit/tensorexpr/kernel.cpp#L958
We can skip bound checks for dis-contiguous tensors for now since there are no such buffer accesses. @ZolotukhinM correct me if I'm wrong.
  
For flattened buffers, we do need to make sure that there are no out of bound buffer accesses.

huiguoo replied:
btw, the size of discontiguous tensors are not computed, and we are using size 0 for `inBuffer`. This will trigger an error in your bound check. I am fixing it in this PR: https://github.com/pytorch/pytorch/pull/69657

davidberard98 replied:
thanks! this addresses my concern about the 1D tensors with 1D flattened buffer

huiguoo commented on test/test_tensorexpr.py:
```diff
@@ -1578,5 +1578,69 @@ def foo(a, b, c):
             exp = traced(a, b, c)
             self.assertEqual(ref, exp)
 
+    def test_1d_strided_within_bounds(self):
```
I think `test/cpp/tensorexpr/test_expr.cpp` is a more appropriate place to add these tests. NNC is implemented in C/C++, and currently its unit tests (files in `test/cpp/tensorexpr`) are mainly written in C++. We are using `test/test_tensorexpr.py` to test the integration of NNC to JIT for op fusions.

davidberard98 commented on torch/csrc/jit/tensorexpr/eval.cpp:
```diff
@@ -672,13 +672,47 @@ class SimpleIREvaluatorImpl : public IRVisitor {
   }
 
+  template <typename T>
+  void check_bounds(T v) {
```
The reason I used templating was for error handling, i.e. 
```
      throw malformed_input(
          "dimensions and indices mismatch in check_bounds", v);
```
and
```
          throw malformed_input(ss.str(), v);
```

Is there another good way to have helpful errors here?

huiguoo replied:
If I understand correctly, we can throw an exception in any code areas without using templating. 
Here's an example (exception thrown in a function):
https://github.com/pytorch/pytorch/blob/3b27304d20f53619344b74e683357424ca3d4029/torch/csrc/jit/tensorexpr/eval.cpp#L407

ZolotukhinM replied:
I think printing `v->buf` would provide enough information for the error message to be helpful, but if you'd like to print Store/Load containing it, we could change `check_bounds` to return bool instead of raising an exception and then raise at the call site.

davidberard98 committed:
Update on "[tensorexpr] check for index out of bounds in ir_eval"


when executing with ir_eval, check for index out of bounds.

Differential Revision: [D32657881](https://our.internmc.facebook.com/intern/diff/D32657881)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/tensorexpr/expr.h:
```diff
@@ -298,6 +306,7 @@ class TORCH_API Buf : public ExprNode<Buf> {
   // qscale_ and qzero_ are used only for quantized dtypes Bufs: kQUInt8, kQInt8
   ExprPtr qscale_;
   ExprPtr qzero_;
+  bool is_flattened_;
```
Generally we should avoid adding new fields to core classes, such as Buf. Is_flattened can be determined from looking at indices/dims, so it's better to have a method than a member field.

The reason for this is that more fields we have, the higher chance is that at some point we end up with an inconsistent state (e.g. multiple indices, multiple dims, but is_flattened=true). Instead of carefully checking that we don't break any invariants there every time we touch one of these fields, we can replace one of them with a method and thus the inconsistent state would never be possible.

ZolotukhinM reviewed (commented):
I haven't looked deeply into the code, but thank you for working on this! It would be very helpful in debugging!

davidberard98 committed:
Update on "[tensorexpr] check for index out of bounds in ir_eval"


when executing with ir_eval, check for index out of bounds.

Differential Revision: [D32657881](https://our.internmc.facebook.com/intern/diff/D32657881)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32657881).

davidberard98 committed:
Update on "[tensorexpr] check for index out of bounds in ir_eval"


when executing with ir_eval, check for index out of bounds.

Differential Revision: [D32657881](https://our.internmc.facebook.com/intern/diff/D32657881)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32657881).

davidberard98 committed:
Update on "[tensorexpr] check for index out of bounds in ir_eval"


when executing with ir_eval, check for index out of bounds.

Differential Revision: [D32657881](https://our.internmc.facebook.com/intern/diff/D32657881)

[ghstack-poisoned]

davidberard98 commented on test/test_tensorexpr_pybind.py:
```diff
@@ -67,31 +67,6 @@ def test_external_calls(self):
         codegen.call([tA, tB, tC])
         torch.testing.assert_close(torch.matmul(tA, tB), tC)
 
-    def test_dynamic_shape(self):
-        dN = te.VarHandle(torch.int32)
-        A = te.BufHandle(torch.float64)
-        B = te.BufHandle(torch.float64)
-
-        def compute(i):
-            return A.load(i) - B.load(i)
-
-        C = te.Compute("C", [dN], compute)
-
-        loopnest = te.LoopNest([C])
-        loopnest.prepare_for_codegen()
-
-        cg = te.construct_codegen("ir_eval", loopnest.simplify(), [A, B, C, dN])
-
-        def test_with_shape(n):
-            tA = torch.randn(n, dtype=torch.double)
-            tB = torch.randn(n, dtype=torch.double)
-            tC = torch.empty(n, dtype=torch.double)
-            cg.call([tA, tB, tC, n])
-            torch.testing.assert_close(tA - tB, tC)
-
-        test_with_shape(8)
-        test_with_shape(31)
-
```
@huiguoo @ZolotukhinM Do you think it's reasonable to remove this test? Or should we use an alternative solution in order to keep this test, e.g. adding a special case for {dims.size() == 0 && indices.size() == 1}

Context: this test creates `BufHandle`s with 0 dimensions and then tests on inputs of different sizes, so it's incompatible with out-of-bounds checks. Note that the 2d version of this test already fails, due to a check in `flatten_index` (since if `indices.size() > strides.size()`, we can't compute the flattened index)

ZolotukhinM replied:
I think we can fix the test:
```diff
diff --git a/test/test_tensorexpr_pybind.py b/test/test_tensorexpr_pybind.py
index 2a0dbfe293a..09d93f8f711 100644
--- a/test/test_tensorexpr_pybind.py
+++ b/test/test_tensorexpr_pybind.py
@@ -69,8 +69,8 @@ class TestTensorExprPyBind(JitTestCase):

     def test_dynamic_shape(self):
         dN = te.VarHandle(torch.int32)
-        A = te.BufHandle(torch.float64)
-        B = te.BufHandle(torch.float64)
+        A = te.BufHandle("A", [dN], torch.float64)
+        B = te.BufHandle("B", [dN], torch.float64)

         def compute(i):
             return A.load(i) - B.load(i)
```
Will it pass after this edit?

ZolotukhinM replied:
> Note that the 2d version of this test already fails

Where is the 2d version?

davidberard98 replied:
This is the test I wrote for 2D:

```python
    def test_dynamic_shape_2d(self):
        zero = te.ExprHandle.int(0)
        dM = te.VarHandle(torch.int32)
        dN = te.VarHandle(torch.int32)
        i = te.VarHandle(torch.int32)
        j = te.VarHandle(torch.int32)
        A = te.BufHandle(torch.float64)
        B = te.BufHandle(torch.float64)
        C = te.BufHandle(torch.float64)

        a_val = A.load([i, j])
        b_val = B.load([i, j])
        add_val = a_val + b_val
        body = te.Store.make(C, [i, j], add_val)

        inner = te.For.make(j, zero, dN, body)
        stmt = te.For.make(i, zero, dM, inner)

        cg = te.construct_codegen("ir_eval", stmt, [A, B, C, dN, dM])

        def test_with_shape(n, m):
            tA = torch.randn(n, m, dtype=torch.double)
            tB = torch.randn(n, m, dtype=torch.double)
            tC = torch.empty(n, m, dtype=torch.double)
            cg.call([tA, tB, tC, n, m])
            torch.testing.assert_close(tA + tB, tC)

        test_with_shape(2, 4)
        test_with_shape(6, 3)
```

ZolotukhinM replied:
Yeah, we just need to specify dimensions for the buffers we create ([dN, dM]), you can do that in the 2d version as well. Previously the test was not rigorous enough, but we got away with that because we didn't have the out-of-bounds check. Now we need to be more careful and consistent with how we define buffers.

Also, you don't need to create IR manually, you could use the same `Compute` API as in 1-d case - this would probably let you make the test slightly more concise.

davidberard98 committed:
Update on "[tensorexpr] check for index out of bounds in ir_eval"


when executing with ir_eval, check for index out of bounds.

Differential Revision: [D32657881](https://our.internmc.facebook.com/intern/diff/D32657881)

[ghstack-poisoned]

davidberard98 committed:
Update on "[tensorexpr] check for index out of bounds in ir_eval"


when executing with ir_eval, check for index out of bounds.

Differential Revision: [D32657881](https://our.internmc.facebook.com/intern/diff/D32657881)

[ghstack-poisoned]

davidberard98 committed:
Update on "[tensorexpr] check for index out of bounds in ir_eval"


when executing with ir_eval, check for index out of bounds.

Differential Revision: [D32657881](https://our.internmc.facebook.com/intern/diff/D32657881)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32657881).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/eval.cpp:
```diff
@@ -672,13 +672,75 @@ class SimpleIREvaluatorImpl : public IRVisitor {
+          ExprPtr buf_size_expr = immLike(dims[0], 1);
+          ExprPtr negative_one = immLike(dims[0], -1);
+          for (const auto& i : c10::irange(dims.size())) {
+            buf_size_expr = alloc<Add>(
+                buf_size_expr,
+                alloc<Mul>(
+                    alloc<Add>(dims[i], negative_one), buf->strides()[i]));
```
Nit: not critical, but this portion might be more readable if we wrapped ExprPtrs with ExprHandles.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/eval.cpp:
```diff
@@ -672,13 +672,75 @@ class SimpleIREvaluatorImpl : public IRVisitor {
+        }
+        indices[0]->accept(this);
+        const auto& index_values = indexVec(value());
+        for (auto& j : index_values) {
+          if (j < 0 || j >= buf_size) {
+            check_bounds_throw(j, buf_size, buf);
+          }
```
Hm, shouldn't we be comparing each index vs the sizes of the corresponding dim? I.e. if there is more than one index in index_values, then we shouldn't be comparing it vs flattened buf size. Am I missing something here?

davidberard98 replied:
This is the case where there the buffer is (n>1)-dimensional, but the index is flattened (i.e. this is the if block for when `dims.size() != indices.size()` and `indices.size() == 1`).

As for why there are multiple indices - I based this off of the implementation of `visit(LoadPtr v)` below, which looks to me like it handles indices as if they can be vectors of indices:

```cpp
    ExprPtr flat_idx =
        flatten_index(v->buf()->dims(), v->indices(), v->buf()->strides());
    flat_idx->accept(this);
    auto index = indexVec(value());
```

So, I'm handling each `index[i]` as if it is a vector of indices. Is this the right way to go about this?

ZolotukhinM replied:
I suspect that this vector should always have just one element. Maybe we can assert that (and in the Load too)? But feel free to keep your implementation as-is since indeed that's how Load-visitor is implemented, I don't want us to go into this rabbit hole now :)

davidberard98 replied:
It looks like sometimes it needs to be handled as a vector (#69961). For example, [Expr.VectorAdd01](https://github.com/pytorch/pytorch/blob/master/test/cpp/tensorexpr/test_expr.cpp#L173-L181).

ZolotukhinM replied:
Ah, right, it makes sense when a load/store is vectorized. Thanks for checking this!

ZolotukhinM commented on test/cpp/tensorexpr/test_expr.cpp:
```diff
@@ -556,6 +556,22 @@ TEST(Expr, DynamicShapeAdd) {
+
+  PaddedBuffer<int> data(20);
+
+  EXPECT_ANY_THROW(SimpleIREvaluator(stmt, {X})(data));
```
Could you please add a couple of more tests to verify N-d cases and N-d + flattened index?

davidberard98 committed:
Update on "[tensorexpr] check for index out of bounds in ir_eval"


when executing with ir_eval, check for index out of bounds.

Differential Revision: [D32657881](https://our.internmc.facebook.com/intern/diff/D32657881)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32657881).

davidberard98 committed:
Update on "[tensorexpr] check for index out of bounds in ir_eval"


when executing with ir_eval, check for index out of bounds.

Differential Revision: [D32657881](https://our.internmc.facebook.com/intern/diff/D32657881)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32657881).

davidberard98 committed:
Update on "[tensorexpr] check for index out of bounds in ir_eval"


when executing with ir_eval, check for index out of bounds.

Differential Revision: [D32657881](https://our.internmc.facebook.com/intern/diff/D32657881)

[ghstack-poisoned]

ZolotukhinM commented on test/cpp/tensorexpr/test_expr.cpp:
```diff
@@ -556,6 +556,65 @@ TEST(Expr, DynamicShapeAdd) {
+    BufHandle X("X", {N, M}, kInt);
+
+    auto body = Store::make(X, {i, j}, i);
+    auto inner = For::make(j, start, stopInner, body);
```
Nit: we don't need to create ExprHandle for immediates, I think the default conversion would work as well, so we can write this as:
```cpp
auto inner = For::make(j, 0, 15, body);
```

Similarly, when we declare `X` we can write:
```cpp
BufHandle X("X", {sizes.first, sizes.second}, kInt);
```

Generally, everything with "Handle" is supposed to be user-friendly and have operators overloaded (e.g. we can write `i + 5` to generate an `Add` expression handle), conversions from default types defined, etc. It is when we work with expr pointers we have to be explicit in everything we do.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/eval.cpp:
```diff
@@ -672,13 +672,74 @@ class SimpleIREvaluatorImpl : public IRVisitor {
+    const std::vector<ExprPtr>& dims = buf->dims();
+    if (dims.size() != indices.size()) {
+      // indices are flattened, but not buffer
+      if (indices.size() == 1) {
```
Nit: From styling perspective it might be better to use assert here instead of if - this would allow us to have 1 less nesting level in the code.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/eval.cpp:
```diff
@@ -672,13 +672,74 @@ class SimpleIREvaluatorImpl : public IRVisitor {
+          ExprHandle negative_one = ExprHandle(immLike(dims[0], -1));
+          for (const auto& i : c10::irange(dims.size())) {
+            buf_size_expr = buf_size_expr +
+                ((negative_one + ExprHandle(dims[i])) *
```
```suggestion
                ((ExprHandle(dims[i]) - 1) *
```

ZolotukhinM commented on torch/csrc/jit/tensorexpr/eval.cpp:
```diff
@@ -672,13 +672,74 @@ class SimpleIREvaluatorImpl : public IRVisitor {
+    }
+    for (const auto& i : c10::irange(dims.size())) {
+      auto opt_dim = intValue(dims[i]);
+      if (!opt_dim) {
```
When can this happen? If it's for size 0, hasn't this been fixed now?

ZolotukhinM reviewed (approved):
Looks good to me!

huiguoo commented:
👍
